ethereum/abi.py

```diff
@@ -1,104 +1,210 @@
+            ]
+
+            signature = [
+                (element['type'], element['name'])
+                for element in description['inputs']
+            ]
+
+            # type can be omitted, defaulting to function
+            if description.get('type', 'function') == 'function':
+                normalized_name = _normalize_name(description['name'])
+
+                decode_types = [
+                    element['type']
+                    for element in description['outputs']
+                ]
+
+                self.function_data[normalized_name] = {
+                    'prefix': method_id(normalized_name, encode_types),
+                    'encode_types': encode_types,
+                    'decode_types': decode_types,
+                    'is_constant': description.get('constant', False),
+                    'signature': signature,
                 }
-            elif sig_item['type'] == 'event':
-                indexed = [f['indexed'] for f in sig_item['inputs']]
-                names = [f['name'] for f in sig_item['inputs']]
-                self.event_data[event_id(name, encode_types)] = {
-                    "types": encode_types,
-                    "name": name,
-                    "names": names,
-                    "indexed": indexed,
-                    "anonymous": sig_item.get('anonymous', False)
+
+            elif description['type'] == 'event':
+                normalized_name = _normalize_name(description['name'])
+
+                indexed = [
+                    element['indexed']
+                    for element in description['inputs']
+                ]
+                names = [
+                    element['name']
+                    for element in description['inputs']
+                ]
+                self.event_data[event_id(normalized_name, encode_types)] = {
+                    'types': encode_types,
+                    'name': normalized_name,
+                    'names': names,
+                    'indexed': indexed,
+                    'anonymous': description.get('anonymous', False),
                 }
 
-    def encode(self, name, args):
-        fdata = self.function_data[name]
-        o = zpad(encode_int(fdata['prefix']), 4) + \
-            encode_abi(fdata['encode_types'], args)
-        return o
+            elif description['type'] == 'constructor':
+                if self.constructor_data is not None:
+                    raise ValueError('Only one constructor is supported.')
 
-    def decode(self, name, data):
-        # print 'out', utils.encode_hex(data)
-        fdata = self.function_data[name]
-        if fdata['is_unknown_type']:
-            o = [utils.to_signed(utils.big_endian_to_int(data[i:i + 32]))
-                 for i in range(0, len(data), 32)]
-            return [0 if not o else o[0] if len(o) == 1 else o]
-        o = decode_abi(fdata['decode_types'], data)
-        return o
+                self.constructor_data = {
```

Comment:
please add a test for using `contructor_data` -- for documentation and test purposes :)